columnphysics/icepack_therm_vertical.F90

```diff
@@ -2381,178 +2382,92 @@ subroutine icepack_step_therm1(dt, ncat, nilyr, nslyr,    &
-         allocate(l_isosno(size(isosno,dim=1),size(isosno,dim=2)))
-         l_isosno     = isosno
-      else
-         allocate(l_isosno(1,1))
-         l_isosno     = c0
-      endif
-
-      if (present(isoice)    ) then
-         allocate(l_isoice(size(isoice,dim=1),size(isoice,dim=2)))
-         l_isoice     = isoice
-      else
-         allocate(l_isoice(1,1))
-         l_isoice     = c0
-      endif
-
-      if (present(Qa_iso)    ) then
-         allocate(l_Qa_iso(size(Qa_iso)))
-         l_Qa_iso     = Qa_iso
-      else
-         allocate(l_Qa_iso(1))
-         l_Qa_iso     = c0
-      endif
-
-      if (present(Qref_iso)    ) then
-         allocate(l_Qref_iso(size(Qref_iso)))
-         l_Qref_iso     = Qref_iso
-      else
-         allocate(l_Qref_iso(1))
-         l_Qref_iso     = c0
-      endif
-
-      if (present(fiso_atm)  ) then
-         allocate(l_fiso_atm(size(fiso_atm)))
-         l_fiso_atm = fiso_atm
-      else
-         allocate(l_fiso_atm(1))
-         l_fiso_atm   = c0
-      endif
-
-      if (present(fiso_ocn)  ) then
-         allocate(l_fiso_ocn(size(fiso_ocn)))
-         l_fiso_ocn = fiso_ocn
-      else
-         allocate(l_fiso_ocn(1))
-         l_fiso_ocn   = c0
-      endif
-
-      if (present(fiso_evap)  ) then
-         allocate(l_fiso_evap(size(fiso_evap)))
-         l_fiso_evap = fiso_evap
-      else
-         allocate(l_fiso_evap(1))
-         l_fiso_evap   = c0
+      if (argcheck == 'always' .or. (argcheck == 'first' .and. first_call)) then
+         if (tr_iso) then
+            if (.not.(present(isosno)   .and. present(isoice)     .and. &
+                      present(fiso_atm) .and. present(fiso_ocn)   .and. &
+                      present(fiso_evap).and.                           &
+                      present(Qa_iso)   .and. present(Qref_iso)   .and. &
+                      present(HDO_ocn)  .and. present(H2_16O_ocn) .and. &
+                      present(H2_18O_ocn))) then
+               call icepack_warnings_add(subname//' error in iso arguments, tr_iso=T')
+               call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+               return
+            endif
+         endif
+         if (snwgrain) then
+            if (.not.(present(rsnwn)  .and.                  &
+                      present(smicen) .and. present(smliqn))) then
+               call icepack_warnings_add(subname//' error in snw arguments, snwgrain=T')
+               call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+               return
+            endif
+         endif
+         if ((present(fswthru_vdr) .and. .not.present(fswthrun_vdr)) .or. &
+             (present(fswthru_vdf) .and. .not.present(fswthrun_vdf)) .or. &
+             (present(fswthru_idr) .and. .not.present(fswthrun_idr)) .or. &
+             (present(fswthru_idf) .and. .not.present(fswthrun_idf))) then
+            call icepack_warnings_add(subname//' error in fswthru [iv]d[rf] arguments')
```

Comment:
@apcraig @dabail10  Is this the logic we want here?  Often it's fine to have category-aggregated variables and not keep the category-specific ones.

Reply:
It's really the underlying implementation.  The implementation in Icepack with this update will always computed the aggregated variables if they are passed, and that requires the category specific arguments too.  We could change the implementation so both have to be passed for the category aggregate to be computed.  The optional argument checking is determined to a large degree on the underlying implementation.